### PR TITLE
Page caching and expiration for components.json

### PR DESCRIPTION
**Pull request by joshjordan**

### CODE REVIEW

**Comment by joshjordan** [comment]
@hut8 I generated a set of `Component`s and `Version`s that puts my locally generated file at around 1.4MB (very similar to production). This takes about 6 seconds to serve with our current strategy, entirely because Rails is implicitly calling `to_json` on the result of the cache call. It just wasnt' structured properly.

We could've cached the result of `to_json` and gotten down to about 500ms for my test set. However, this work caches `components.json` as a statically served asset. In development mode, this gets me down to about 440ms on average but I expect that nginx actually serves a large static asset much faster.

File: app/controllers/components_controller.rb [hunk 1]
```diff
@@ -1,13 +1,11 @@
 class ComponentsController < ApplicationController
+  caches_page :index
```
Comment (joshjordan) [inline app/controllers/components_controller.rb]: This is what actually generates `public/components.json`

File: app/models/page_cache_manager.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  include Rails.application.routes.url_helpers
+
+  ACTION_CONTROLLER_METHODS = [:page_cache_directory, :perform_caching, :default_static_extension]
+
+  ACTION_CONTROLLER_METHODS.each do |method|
+    define_singleton_method method do
+      ActionController::Base.send(method)
+    end
+  end
+end
```
Comment (joshjordan) [inline app/models/page_cache_manager.rb]: This class is basically an `ActionController::Caching::Pages` instance, plus the methods it needs off of a Rails controller in order to work. See tests for proof.

File: app/workers/reindex.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def perform(force = false)
     end
 
     # Clear cache for components.json api endpoint
-    Rails.cache.delete('components_json')
+    PageCacheManager.new.expire_page(controller: :components, action: :index, format: :json)
```
Comment (joshjordan) [inline app/workers/reindex.rb]: `ActionController::Caching::Pages` translates this to `components.json`, concatenates it with the default caching path (`public/`), and then deletes it and any `.gz` version that may be present.

File: config/deploy.rb [hunk 4]
```diff
@@ -13,8 +13,10 @@
 set :log_level, :debug
 set :pty, false
 set(:linked_files,
-    fetch(:linked_files, []).push(
-      'config/database.yml', 'config/application.yml'))
+  fetch(:linked_files, []).push(
+    'config/database.yml', 'config/application.yml', 'public/components.json'
```
Comment (joshjordan) [inline config/deploy.rb]: Any reason not to link it?

Comment (hut8) [inline config/deploy.rb]: Not that I can see. I have another issue open to link more things. I'm actually nervous about deploying this without it.

**Pull request merged**